File: appkillermanager/src/main/java/com/thelittlefireman/appkillermanager/utils/SystemUtils.java
package com.thelittlefireman.appkillermanager.utils;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.util.Log;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class SystemUtils {

    public static String getDefaultDebugInformation(){
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("Display_id:").append(Build.DISPLAY);

        stringBuilder.append("MODEL:").append(Build.MODEL);
        stringBuilder.append("MANUFACTURER:").append(Build.MANUFACTURER);
        stringBuilder.append("PRODUCT:").append(Build.PRODUCT);
        return stringBuilder.toString();
    }

    public static String getEmuiRomName() {
        try {
            return SystemUtils.getSystemProperty("ro.build.version.emui");
        } catch (Exception e) {
            return "";
        }
    }
    public static String getApplicationName(Context context) {
        PackageManager packageManager = context.getPackageManager();
        ApplicationInfo applicationInfo = null;
        try {
            applicationInfo = packageManager.getApplicationInfo(context.getApplicationInfo().packageName, 0);
        } catch (final PackageManager.NameNotFoundException e) {
        }
        return (String) (applicationInfo != null ? packageManager.getApplicationLabel(applicationInfo) : "Unknown");
    }

    public static String getMiuiRomName() {
        try {
            return SystemUtils.getSystemProperty("ro.miui.ui.version.name");
        } catch (Exception e) {
            return "";
        }
    }

    private static String getSystemProperty(String propName) {
        String line;
        BufferedReader input = null;
        try {
            Process p = Runtime.getRuntime().exec("getprop " + propName);
            input = new BufferedReader(new InputStreamReader(p.getInputStream()), 1024);
            line = input.readLine();
            input.close();
        } catch (IOException ex) {
            Log.e(SystemUtils.class.getClass().getName(), "Unable to read system property " + propName, ex);
            return null;
        } finally {
            if (input != null) {
                try {
                    input.close();
                } catch (IOException e) {
                    Log.e(SystemUtils.class.getClass().getName(), "Exception while closing InputStream", e);
                }
            }
        }
        return line;
    }
}

File: appkillermanager/src/main/java/com/thelittlefireman/appkillermanager/utils/SystemUtils.kt
package com.thelittlefireman.appkillermanager.utils

import android.annotation.TargetApi
import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.os.Build
import android.os.Process
import android.os.UserManager
import androidx.annotation.RequiresApi
import timber.log.Timber
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

object SystemUtils {
    val defaultDebugInformation: String
        get() = "Display_id: ${Build.DISPLAY}\n" +
                "MODEL: ${Build.MODEL}\n" +
                "MANUFACTURER: ${Build.MANUFACTURER}\n" +
                "PRODUCT: ${Build.PRODUCT}"

    val emuiRomName: String?
        get() = try {
            getSystemProperty("ro.build.version.emui")
        } catch (e: Exception) {
            ""
        }

    fun getApplicationName(context: Context): String {
        val packageManager = context.packageManager
        var applicationInfo: ApplicationInfo? = null
        try {
            applicationInfo =
                packageManager.getApplicationInfo(context.applicationInfo.packageName, 0)
        } catch (e: PackageManager.NameNotFoundException) {
        }
        return (if (applicationInfo != null) packageManager.getApplicationLabel(applicationInfo) else "Unknown") as String
    }

    val miuiRomName: String?
        get() = try {
            getSystemProperty("ro.miui.ui.version.name")
        } catch (e: Exception) {
            ""
        }

    private fun getSystemProperty(propName: String): String? {
        val line: String
        var input: BufferedReader? = null
        try {
            val p = Runtime.getRuntime().exec("getprop $propName")
            input = BufferedReader(InputStreamReader(p.inputStream), 1024)
            line = input.readLine()
            input.close()
        } catch (ex: IOException) {
            Timber.e(ex, "Unable to read system property %s", propName)
            return null
        } finally {
            if (input != null) {
                try {
                    input.close()
                } catch (e: IOException) {
                    Timber.e(e, "Exception while closing InputStream")
                }
            }
        }
        return line
    }
    // INFO http://imsardine.simplbug.com/note/android/adb/commands/am-start.html
    /**
     * Open an Activity by using Application Manager System (prevent from crash permission exception)
     *
     * @param context         current application Context
     * @param packageName     pacakge name of the target application (exemple: com.huawei.systemmanager)
     * @param activityPackage activity name of the target application (exemple: .optimize.process.ProtectActivity)
     */
    @RequiresApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    fun startActivityByAMSystem(context: Context, packageName: String, activityPackage: String) {
        var cmd = "am start -n $packageName/$activityPackage"
        val um = context.getSystemService(Context.USER_SERVICE) as UserManager
        cmd += " --user " + um.getSerialNumberForUser(Process.myUserHandle())
        Runtime.getRuntime().exec(cmd)
    }

    /**
     * Open an Action by using Application Manager System (prevent from crash permission exception)
     *
     * @param context      current application Context
     * @param intentAction action of the target application (exemple: com.huawei.systemmanager)
     */
    @RequiresApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    fun startActionByAMSystem(context: Context, intentAction: String) {
        var cmd = "am start -a $intentAction"
        val um = context.getSystemService(Context.USER_SERVICE) as UserManager
        cmd += " --user " + um.getSerialNumberForUser(Process.myUserHandle())
        Runtime.getRuntime().exec(cmd)
    }
}
